src/V3Mutex.h

```diff
@@ -0,0 +1,164 @@
+///
+/// \file
+/// \brief Verilator Mutex and LockGuard used only in verilator.
+///
+/// This file defines Mutex and LockGuard that is used in verilator
+/// source code. It shouldn't be used by verilated code.
+/// In contrast to VerilatedMutex that is used by verilated code,
+/// this mutex can be configured and disabled when verilation is using
+/// single thread.
+///
+/// This implementation also allows using different base mutex class in
+/// wrapper V3Mutex class.
+///
+//*************************************************************************
+
+#ifndef VERILATOR_V3MUTEX_H_
+#define VERILATOR_V3MUTEX_H_ 1
+
+#include "verilatedos.h"
+
+#include <cstdlib>
+#include <functional>
+#include <mutex>
+
+#define VL_LOCK_SPINS 50000  /// Number of times to spin for a mutex before yielding
+
+// MutexConfig class that allows to configure how mutex and lockgurads behave
+// once configured and locked, it cannot be changed. Configuration and lock needs to be
+// done before starting any additional threads.
+class V3MutexConfig final {
+private:
+    // Allows to disable mutexes and lockguards.
+    // Use carefully as it can cause undefined behavior when used inappropriately.
+    // All mutexes needs to be unlocked.
+    bool m_enable = false;  // Allows locking on mutexes, default don't lock mutexes
+    bool m_lockConfig = false;  // After set, configuration cannot be changed
+
+    V3MutexConfig() = default;
+    ~V3MutexConfig() = default;
+
+public:
+    static V3MutexConfig& s() VL_MT_SAFE {
+        static V3MutexConfig s;
+        return s;
+    }
+
+    // configures class
+    void configure(bool enable) VL_MT_UNSAFE {
+        if (!s().m_lockConfig) {
+            s().m_enable = enable;
+            s().m_lockConfig = true;
+        } else {
+            // requires <iostream>
+            // avoided to reduce compile time
+            // std::cerr << "%Error: V3Mutex already configured." << std::endl;
+            std::abort();
+        }
+    }
+    bool lockConfig() VL_MT_SAFE { return m_lockConfig; }
+    bool enable() VL_MT_SAFE { return m_enable; }
+};
+
+/// Mutex, wrapped to allow -fthread_safety checks
+template <typename T>
+class VL_CAPABILITY("mutex") V3MutexImp final {
+private:
+    T m_mutex;  // Mutex
+
+public:
+    /// Construct mutex (without locking it)
+    V3MutexImp() = default;
+    ~V3MutexImp() = default;
+    const V3MutexImp& operator!() const { return *this; }  // For -fthread_safety
+    /// Acquire/lock mutex
+    void lock() VL_ACQUIRE() VL_MT_SAFE {
+        if (V3MutexConfig::s().enable()) {
+            // Try to acquire the lock by spinning.  If the wait is short,
+            // avoids a trap to the OS plus OS scheduler overhead.
+            if (VL_LIKELY(try_lock())) return;  // Short circuit loop
+            for (int i = 0; i < VL_LOCK_SPINS; ++i) {
+                if (VL_LIKELY(try_lock())) return;
+                VL_CPU_RELAX();
```

Comment:
We should increment some V3Stat yield counter (per stage) when we do this to help performance debug, but we can add that later.